File: leetcode-js/333-largest-bst-subtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var largestBSTSubtree = function(root) {
    const result = solve(root);
    return Math.abs(result.res);
};

function solve(root) {
    if (!root) {
        return {
            res: 0,
            min: Infinity,
            max: -Infinity
        };
    }
    const left = solve(root.left);
    const right = solve(root.right);
    if (left.res < 0 || right.res < 0 || root.val < left.max || root.val > right.min) {
        return {
            res: Math.max(Math.abs(left.res), Math.abs(right.res)) * -1,
            min: 0,
            max: 0
        };
    } else {
        return {
            res: left.res + right.res + 1,
            min: Math.min(root.val, left.min),
            max: Math.max(root.val, right.max)
        };
    }
}

File: leetcode-java/333-largest-bst-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    class Result {
        int size;
        int lower;
        int upper;
        Result(int size, int lower, int upper) {
            this.size = size;
            this.lower = lower;
            this.upper = upper;
        }
    }

    int max = 0;

    public int largestBSTSubtree(TreeNode root) {
        helper(root);
        return max;
    }

    Result helper(TreeNode node) {
        if (node == null) {
            return new Result(0, Integer.MAX_VALUE, Integer.MIN_VALUE);
        }

        Result left = helper(node.left);
        Result right = helper(node.right);

        if (left.size == -1 || right.size == -1 || left.upper >= node.val || right.lower <= node.val) {
            return new Result(-1, 0, 0);
        }

        int size = 1 + left.size + right.size;

        max = Math.max(max, size);

        return new Result(size, Math.min(left.lower, node.val), Math.max(right.upper, node.val));
    }
}
